src/device/simulator.ts

```diff
@@ -0,0 +1,153 @@
+{
+  status: ConnectionStatus = ConnectionStatus.NO_AUTHORIZED_DEVICE;
+  sensors: Record<string, Sensor> = {};
+
+  private messageListener = (event: MessageEvent) => {
+    const iframe = this.iframe();
+    if (!iframe || event.source !== iframe.contentWindow || !event.data.kind) {
+      // Not an event for us.
+      return;
+    }
+
+    switch (event.data.kind) {
+      case "ready": {
+        this.sensors = sensorsById(event);
+        this.emit(EVENT_SENSORS, this.sensors);
+        // For now we're always connected when ready.
+        // In the future we'll disconnect when stopped.
+        this.setStatus(ConnectionStatus.CONNECTED);
+        break;
+      }
+      case "sensor_change": {
+        this.sensors = sensorsById(event);
+        this.emit(EVENT_SENSORS, this.sensors);
+        break;
+      }
+      case "serial_output": {
+        const text = event.data.data;
+        if (typeof text === "string") {
+          this.emit(EVENT_SERIAL_DATA, text);
+        }
+        break;
+      }
+      default: {
+        // Ignore unknown message.
+      }
+    }
+  };
+
+  constructor(private iframe: () => HTMLIFrameElement | null) {
+    super();
+  }
+
+  async initialize(): Promise<void> {
+    window.addEventListener("message", this.messageListener);
+    this.setStatus(ConnectionStatus.NOT_CONNECTED);
+  }
+
+  dispose() {
+    this.removeAllListeners();
+    window.removeEventListener("message", this.messageListener);
+  }
+
+  async connect(): Promise<ConnectionStatus> {
+    this.setStatus(ConnectionStatus.CONNECTED);
+    return this.status;
+  }
+
+  async flash(
+    dataSource: FlashDataSource,
+    options: {
+      partial: boolean;
+      progress: (percentage: number | undefined) => void;
+    }
+  ): Promise<void> {
+    this.postMessage("flash", {
+      filesystem: await dataSource.files(),
+    });
+    options.progress(undefined);
+    this.emit(EVENT_FLASH);
+  }
+
+  async disconnect(): Promise<void> {
+    window.removeEventListener("message", this.messageListener);
+    this.setStatus(ConnectionStatus.NOT_CONNECTED);
+  }
+
+  async serialWrite(data: string): Promise<void> {
+    this.postMessage("serial_input", {
+      data,
+    });
+  }
+
+  sensorWrite = async (id: string, value: number): Promise<void> => {
+    this.postMessage("sensor_set", {
+      sensor: id,
+      value,
+    });
+  };
+
+  stop = async (): Promise<void> => {
+    this.postMessage("serial_input", {
+      // Ctrl-C to interrupt.
+      // A specific message would be useful as probably best to clear display etc. here.
+      data: `\x03`,
+    });
+  };
+
+  private setStatus(newStatus: ConnectionStatus) {
+    this.status = newStatus;
+    this.emit(EVENT_STATUS, this.status);
+  }
+
+  clearDevice(): void {
+    this.setStatus(ConnectionStatus.NO_AUTHORIZED_DEVICE);
+  }
+
+  private postMessage(kind: string, data: any): void {
+    const iframe = this.iframe();
+    if (!iframe) {
+      throw new Error("No-one to talk to.");
```

Comment:
Probably more whimsical than is prudent.

Reply:
[`03883e5`](https://github.com/microbit-foundation/python-editor-next/pull/852/commits/03883e55f0d0872a930d4e854b722a43ef75e7fc)